textattack/metrics/quality_metrics/perplexity.py

```diff
@@ -0,0 +1,101 @@
+import re
+
+import torch
+import tqdm
+from transformers import AutoTokenizer, GPT2LMHeadModel, GPT2Tokenizer
+
+from textattack.attack_results import FailedAttackResult, SkippedAttackResult
+from textattack.metrics import Metric
+
+
+class Perplexity(Metric):
+    """Calculates average Perplexity on all successfull attacks using a pre-trained small GPT-2 model
+
+    Args:
+    results (:obj::`list`:class:`~textattack.goal_function_results.GoalFunctionResult`):
+                    Attack results for each instance in dataset
+    """
+
+    def __init__(self, results):
+        self.results = results
+        self.all_metrics = {}
+        self.original_candidates = []
+        self.successful_candidates = []
+        self.ppl_model = GPT2LMHeadModel.from_pretrained("gpt2")
+        if torch.cuda.is_available():
+            self.ppl_model.cuda()
+        self.ppl_tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
+        self.ppl_model.eval()
+
+        self.original_candidates_ppl = []
+        self.successful_candidates_ppl = []
+
+    def calculate(self):
+
+        for i, result in enumerate(self.results):
+            if isinstance(result, FailedAttackResult):
+                continue
+            elif isinstance(result, SkippedAttackResult):
+                continue
+            else:
+                self.original_candidates.append(
+                    result.original_result.attacked_text.text.lower()
+                )
+                self.successful_candidates.append(
+                    result.perturbed_result.attacked_text.text.lower()
+                )
+
+        self.all_metrics["avg_original_perplexity"] = round(
+            self.calc_ppl(self.original_candidates)[0], 2
+        )
+        self.all_metrics["avg_attack_perplexity"] = round(
+            self.calc_ppl(self.successful_candidates)[0], 2
+        )
+
+        return self.all_metrics
+
+    def calc_ppl(self, texts):
+        eval_loss = 0
+        ppl_losses = []
+        nb_eval_steps = 0
+
+        with torch.no_grad():
+            for text in texts:
+                text = self.process_string(text)
+                input_ids = torch.tensor(
```

Comment:
In this case, it seems like we're just truncating the end of the text if it is longer than 1024 (or whatever the max length of GPT2 is). There's an alternative method where we calculate perplexity in strides. Look at this for more information: https://huggingface.co/transformers/perplexity.html

Personally, I think this is a fairer way to calculate perplexity.